### PR TITLE
Add raw stake and median to shard state and roster

### PR DESCRIPTION
**Pull request by gupadhyaya**
reuse all the code from `GetMedianRawStakeSnapshot`, but only work with validator snapshot to recompute the results of previous election. the format of the rpc result is same as `GetMedianRawStakeSnapshot`.

### CODE REVIEW

File: shard/committee/assignment.go [hunk 1]
```diff
@@ -123,6 +142,7 @@ func NewEPoSRound(stakedReader StakingCandidatesReader) (
 
 func prepareOrders(
 	stakedReader StakingCandidatesReader,
+	snapshotOnly bool,
```
Comment (fxfactorial) [inline shard/committee/assignment.go:145]: no bool parameters please. Just make two functions for the branch. prepareOrdersSnapshot, prepareOrders

**Comment by rlan35** [comment]
This is not 100% correct. Because when you recompute the election, you are doing it on the latest list of validators (even though you use the snapshot data).

The correct and simpler way is just in the GetSuperCommitte. Add a field of raw stake and compute the raw stake based on snapshot data. So you don't need to modify EPoS compute functions. Just add some data in RPC results.

File: shard/shard_state.go [hunk 2]
```diff
@@ -33,8 +33,9 @@ const (
 
 // State is the collection of all committees
 type State struct {
-	Epoch  *big.Int    `json:"epoch"`
-	Shards []Committee `json:"shards"`
+	Epoch       *big.Int    `json:"epoch"`
+	Shards      []Committee `json:"shards"`
+	MedianStake numeric.Dec `json:"epos-median-stake" rlp:"nil"`
```
Comment (fxfactorial) [inline shard/shard_state.go:38]: we should not change this field for this

File: shard/committee/assignment.go [hunk 3]
```diff
@@ -183,6 +202,61 @@ func prepareOrders(
...
+func prepareOrdersSnapshot(
+	stakedReader StakingCandidatesReader,
+	candidates []common.Address,
+) (map[common.Address]*effective.SlotOrder, error) {
+	blsKeys := map[shard.BLSPublicKey]struct{}{}
+	essentials := map[common.Address]*effective.SlotOrder{}
+	totalStaked, tempZero := big.NewInt(0), numeric.ZeroDec()
+
+	for i := range candidates {
+		snapshot, err := stakedReader.ReadValidatorSnapshot(
```
Comment (rlan35) [inline shard/committee/assignment.go:214]: Should use the right snapshot. Use ReadValidatorSnapshotAtEpoch(...

File: hmy/api_backend.go [hunk 4]
```diff
@@ -587,6 +587,18 @@ func (b *APIBackend) GetSuperCommittees() (*quorum.Transition, error) {
...
 		quorum.NewRegistry(stakedSlotsNow)
 
+	then.MedianStake, then.SlotPurchases =
+		committee.ComputeMedianRawStakeForShardCommittee(
+			b.hmy.BlockChain(),
+			prevCommittee.StakedValidators().Addrs,
+		)
+
+	now.MedianStake, now.SlotPurchases =
+		committee.ComputeMedianRawStakeForShardCommittee(
```
Comment (rlan35) [inline hmy/api_backend.go:597]: Actually no need to be this complicated. Just add a field in the final result:

 {
                            "is-harmony-slot": false,
                            "earning-account": "one1x8kgyyqgslzrm7jrlj4m7rpg285etdywjk7gez",
                            "bls-public-key": "5e8635085788492a87d3eab69732ae72a55087e95c210ad20a726dd169a69b5792cb94616f6eccc4ea0a54a4a77ca88c",
                            "voting-power-unnormalized": "0.026785021562679984",
                            "voting-power-%": "0.008571206900057595",
                            "effective-stake": "5920773850000000000000000.000000000000000000"
                            "raw-stake": NEW_FIELD,
                        },

The raw-stake is just validatorSnap.TotalStake / size(validatorSnap.BLSKeys). 

No need to add new func like: ComputeMedianRawStakeForShardCommittee. Just loop through all the validators in the decider result and for each validator address, fetch the snapshot and compute the raw-stake and put it in the result.

File: shard/committee/assignment.go (hunk 3 above)
Comment (rlan35) [reply shard/committee/assignment.go:214]: And no need to add new function just for the single RPC.

**Comment by gupadhyaya** [comment]
updated https://github.com/harmony-one/harmony/pull/2880

**Pull request closed without merging**